compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt

```diff
@@ -92,112 +84,107 @@ internal class ModuleMerger(
-          .annotationOrNull(contributesMultibindingFqName)
-        val contributesSubcomponentAnnotation = classDescriptorForExclusion
-          .annotationOrNull(contributesSubcomponentFqName)
+        val contributesToAnnotation = excludedClass.annotationOrNull(contributesToFqName)
+        val contributesBindingAnnotation = excludedClass.annotationOrNull(contributesBindingFqName)
+        val contributesMultibindingAnnotation =
+          excludedClass.annotationOrNull(contributesMultibindingFqName)
+        val contributesSubcomponentAnnotation =
+          excludedClass.annotationOrNull(contributesSubcomponentFqName)
 
         // Verify that the replaced classes use the same scope.
-        val scopeOfExclusion = contributesToAnnotation?.scope(module)
-          ?: contributesBindingAnnotation?.scope(module)
-          ?: contributesMultibindingAnnotation?.scope(module)
-          ?: contributesSubcomponentAnnotation?.parentScope(module)
-          ?: throw AnvilCompilationException(
-            thisDescriptor,
-            "Could not determine the scope of the excluded class " +
-              "${classDescriptorForExclusion.fqNameSafe}."
+        val scopeOfExclusion = contributesToAnnotation?.scopeOrNull()
+          ?: contributesBindingAnnotation?.scopeOrNull()
+          ?: contributesMultibindingAnnotation?.scopeOrNull()
+          ?: contributesSubcomponentAnnotation?.parentScope()
+          ?: throw AnvilCompilationExceptionClassReference(
+            message = "Could not determine the scope of the excluded class " +
+              "${excludedClass.fqName}.",
+            classReference = excludedClass
           )
 
-        if (scopeOfExclusion.fqNameSafe != scopeFqName) {
-          throw AnvilCompilationException(
-            thisDescriptor,
-            "${thisDescriptor.fqNameSafe} with scope $scopeFqName wants to exclude " +
-              "${classDescriptorForExclusion.fqNameSafe} with scope " +
-              "${scopeOfExclusion.fqNameSafe}. The exclusion must use the same scope."
+        if (scopeOfExclusion.fqName != scopeFqName) {
+          throw AnvilCompilationExceptionClassReference(
+            classReference = clazz,
+            message = "${clazz.fqName} with scope $scopeFqName wants to exclude " +
+              "${excludedClass.fqName} with scope " +
+              "${scopeOfExclusion.fqName}. The exclusion must use the same scope."
           )
         }
 
         argumentType.asmType(codegen.typeMapper)
-      }
-      ?: emptyList()
+      } ?: emptyList()
 
     val replacedModules = modules
-      .filter { (classDescriptor, _) ->
+      .filter { (contributedClass, _) ->
         // Ignore replaced modules or bindings specified by excluded modules.
-        classDescriptor.defaultType.asmType(codegen.typeMapper) !in excludedModules
+        contributedClass.clazz.defaultType.asmType(codegen.typeMapper) !in excludedModules
```

Comment:
Is this conversion necessary? Can't `excludedModules` hold instances of `ClassReference`?